src/commands.rs

```diff
@@ -1243,22 +1245,25 @@ fn cmd_config(
     ui.request_pager();
     let settings = command.settings();
     match subcommand {
-        ConfigSubcommand::List { name } => {
-            let raw_values = match name {
-                Some(name) => {
-                    settings
-                        .config()
-                        .get::<config::Value>(name)
-                        .map_err(|e| match e {
-                            config::ConfigError::NotFound { .. } => {
-                                user_error("key not found in config")
-                            }
-                            _ => e.into(),
-                        })?
+        ConfigSubcommand::List {
+            name,
+            include_defaults,
+        } => {
+            let values = command
+                .resolved_config_values(name.as_ref().map(Deref::deref), *include_defaults)?;
```

Comment:
The ref manipulations here seem awkward. Is there a cleaner way to handle the refs/types?

Reply:
You can use `.as_deref()`, but I think `resolved_config_values()` does too much things.

Maybe we can instead add `layered_configs.get_source(key: &str) -> Option<ConfigSource>` which looks up `sources()` in reverse order and returns the first match. Since `write_config_entry()` decomposes intermediate table recursively, we can get the source of the leaf value there, and omit if it was `ConfigSource::Default`.

It wouldn't be as efficient as your implementation, but the whole config entries wouldn't be big.

Reply:
Hmm, I agree it did too many things, but having entirely separate implementations for getting the value and the source feels icky to me, not just because of performance but because it needs to duplicate some lookup logic and there could be discrepancies with how it handles precedence.

I did move the `--include-default` handling entirely out of the helper, though, and I suspect I'll be able to simplify down the logic for overridden values a little further eventually.

Reply:
> there could be discrepancies with how it handles precedence.

That's unavoidable since we rely on `Config::builder()` to merge configs to be actually used, right? We can make sure that each `(value, source)` pair in `resolved_config_values()` is consistent, but it doesn't mean the value would actually be picked by `user_settings.config().get()`.

Reply:
Yes, but feels slightly more bizarre to me if discrepancies could lead to reporting the right value with the wrong source. Obviously most of this logic ultimately ought to be handled by the config crate, but for now I'd really prefer to not have such piecemeal logic for values vs sources or it's gonna get ugly fast trying to support `config list --include-overridden --show-origin`.

Anyway, I realized I should be able to simplify way down by splitting out the operations doing key name lookup. It'll take me a bit to rework the code but I think it'll look much much cleaner. Stay tuned...

Reply:
> Obviously most of this logic ultimately ought to be handled by the config crate, [...]

Yep, so it would be nice if we can come up with a simple workaround for the moment. I would prioritize "right value" than "wrong source", but I'm fine with any solution that looks obviously correct.

Reply:
K, now the code's separated out for resolving the annotated config values vs. filtering the key names. The logic is kinda complex, and hopefully most of it's temporary as the config crate adds more support for these kinds of operations, but it does alright at providing the desired behavior.